Activate mcsolve tests on Windows in QuTiP 5

hodgestar opened the pull request:
**Description**
Activate the mcsolve tests in QuTiP 5.

Changes:
- [x] Swap parallel_map over to ProcessPoolExecutor.
- [x] Switch ``mcsolve`` to the new parallel_map. 
- [x] Reactivated mcsolve tests on Windows. 

**Related issues or PRs**
- Continues work from #1853 and #1854 in v4
- #1202 (maybe already resolved in QuTiP 5)
- #1190 (hopefully already resolved in QuTiP 5)

hodgestar commented:
A bit surprisingly, all tests now pass on Windows on QuTiP 5.

I'm not 100% sure the parallel map change fixed everything, or if everything was mostly working before. I would like to keep the parallel map change regardless, because it removes some of the management of making multiprocessing work and leaves it with the core Python developers.

Ericgig reviewed (commented):
`ProcessPoolExecutor` being enough to fix most windows tests is surprising.
Now we just need to speed up correlation tests.

hodgestar commented on qutip/solver/parallel.py:
```diff
@@ -145,39 +146,41 @@ def parallel_map(task, values, task_args=None, task_kwargs=None,
+        ctx_kw = {}
 
+    os.environ['QUTIP_IN_PARALLEL'] = 'TRUE'
+    try:
```
I hadn't realized the intention of catching the error. Let me see if I can fix things up so that results are retained properly.

Maybe I can also fix `ProcessPoolExecutor` waiting too long.

hodgestar commented on qutip/solver/parallel.py:
```diff
@@ -145,39 +146,41 @@ def parallel_map(task, values, task_args=None, task_kwargs=None,
 
+    os.environ['QUTIP_IN_PARALLEL'] = 'TRUE'
+    try:
+        with ProcessPoolExecutor(
+            max_workers=map_kw['num_cpus'], **ctx_kw,
+        ) as executor:
+            result_futures = []
+            for value in values:
+                future = executor.submit(
+                    task, *((value,) + task_args), **task_kwargs,
+                )
+                result_futures.append(future)
+
+            results = []
+            for f in result_futures:
+                remaining_time = min(end_time - time.time(), job_time)
+                result = f.result(timeout=remaining_time)
+                if reduce_func is not None:
+                    reduce_func(result)
+                else:
+                    results.append(result)
+                progress_bar.update()
```
Thanks, I will try a variation on these!

hodgestar commented:
Interesting discussion about the relationship between OpenMP and multiprocessing -- https://scikit-learn.org/stable/faq.html#why-do-i-sometime-get-a-crash-freeze-with-n-jobs-1-under-osx-or-linux (in case this is relevant when bringing OpenMP back).

hodgestar commented:
@Ericgig I implemented a version of your suggestion. Let's see how it goes on this test run.

Other notes:
- I switched the fork method to `forkserver` for Linux, as that should play better with threads.
- I left the correlation tests in for now, but will remove them again if they still timeout.

Perhaps we should set some timeouts in the correlation function tests as a way to prevent the whole test run from dying.

hodgestar commented:
All tests passing again -- ready for another review!

Ericgig commented on qutip/solver/parallel.py:
```diff
@@ -9,13 +9,18 @@
 
 if sys.platform == 'darwin':
-    Pool = multiprocessing.get_context('fork').Pool
+    mp_context = multiprocessing.get_context('fork')
```
For mac, we used `fork` since `spawn` would fail pickling, but it may work with `concurrent`. We should try it eventually.

hodgestar replied:
I *think* spawn would actually work (or at least it works for our Windows tests). Either `spawn` or `forkserver` do require more complex data sharing since things are no longer in the same process. I think moving data around is what kills performance (Linux correlation tests timeout with `forkserver` too).

Ericgig commented on qutip/tests/solver/test_parallel.py:
```diff
@@ -68,4 +68,4 @@ def test_map_accumulator(map, num_cpus):
 
     map(_func2, x, args, kwargs, reduce_func=y2.append, map_kw=map_kw)
-    assert ((np.array(y1) == np.array(y2)).all())
+    assert ((np.array(sorted(y1)) == np.array(sorted(y2))).all())
```
Is this still needed? Isn't this redundant with the `future._i`?

hodgestar replied:
It's needed when supplying a `reduce_func` to `parallel_map`. We sort the results list, but the `reduce_func` is called in whatever order the tasks complete.

Ericgig reviewed (approved):
Seems to be working fine.
Timeout work as expected and `KeyboardInterrupt` stop the jobs.
It continues when tasks raise exceptions, returning the results with `None`s mixed in.

It's good to know that correlation is only slow but works.
